The invocation of CiliumIdentity changes is triggered after a random delay when modifying the namespace label.

orange30 opened the pull request:
Currently, in clusters where Cilium is deployed, changing the namespace label will instantly generate a number of ciliumidentity when the namespace have many pod who's label is different.
After these ciliumidentity events are pushed to the API server, they are fully distributed to each node.
These ciliumidentity changes will result in a large number of ciliumendpoint update events.
Therefore, when there are many pods who's lable is different in a namespace and the cluster has a large number of nodes, changing the namespace label can easily cause significant pressure on the API server and, under extreme circumstances, may lead to the API server crashing.

1. In this patch, when modifying the namespace label, we delay the invocation of the CiliumIdentity change event by a random amount of time to alleviate pressure on the API server.
2. The maximum delay time can be configured via the cilium-config ConfigMap.
3. Modifying the namespace label is a relatively infrequent operation, so we believe that a delay should be acceptable in most cases.

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #38030

```release-note
A new option `--identity-max-jitter` adds a random delay to when Cilium reacts to namespace label updates. This jitter is intended to improve kube-apiserver resource usage at high scale.
```

squeed commented on pkg/k8s/watchers/namespace.go:
```diff
@@ -122,6 +125,13 @@ func getNamespaceLabels(ns *slim_corev1.Namespace) labels.Labels {
 	return labels.Map2Labels(labelMap, labels.LabelSourceK8s)
 }
 
+// randomDuration 在 0 到 maxDuration 之间生成一个随机的 time.Duration
```
nit: please comment in English :-)

orange30 replied:
I will fix it.

squeed commented on pkg/controller/controller.go:
```diff
@@ -105,6 +105,8 @@ type ControllerParams struct {
 
 	Context context.Context
+
+	DelayUpdate time.Duration
```
Suggestion: call this jitter? And do please add a docblock describing the purpose.

orange30 replied:
I think the exact meaning of this parameter is：
“// The interval for which this update needs to be delayed”

So, for this parameter name, between "jitter" and "delay", I think maybe we shold use "delay" better.

joestringer commented on pkg/controller/manager.go:
```diff
@@ -108,7 +121,17 @@ func (m *Manager) createControllerLocked(name string, params ControllerParams) *
+		default:
+		}
+	} else {
+		go delayUpdate(ctrl, params)
```
What happens if another update arrives during the delay? I think this could revert another change that was already applied?

orange30 replied:
For the case runIdentityResolver->func (e *Endpoint) identityLabelsChanged(ctx context.Context) :
If the second random delay is smaller then the first  random delay, then the second identityLabelsChanged will be executed earlier.  When the first identityLabelsChanged is called, it will returned for obsolete identity revision.

         myChangeRev := e.identityRevision; 
        // Since we unlocked the endpoint and re-locked, the label update may already be obsolete
	if e.identityResolutionIsObsolete(myChangeRev) {
		e.unlock()
		elog.Debug("Endpoint identity has changed, aborting resolution routine in favour of new one")
		return false, nil
	}

joestringer commented on pkg/controller/manager.go:
```diff
@@ -52,6 +52,15 @@ func (m *Manager) UpdateController(name string, params ControllerParams) {
 	m.updateController(name, params)
 }
 
+func delayUpdate(ctrl *managedController, params ControllerParams) {
```
In the current code, controller updates occur while holding `m.mutex`, but the goroutine doesn't make the change under the mutex. We should probably be consistent about how the changes are applied.

orange30 replied:
I will fix it.

joestringer commented on pkg/controller/controller.go:
```diff
@@ -239,93 +241,100 @@ func (c *controller) GetLastErrorTimestamp() time.Time {
+	var params ControllerParams
 	errorRetries := 1
+	runFunc := false
+	interval := 10 * time.Minute
```
Why does this need to start at 10 minutes?

orange30 replied:
https://github.com/cilium/cilium/pull/38031/files#r1991017071
As we analyzed in this link, interval is only meaningful after the "update" event is executed.
I think it might be more reasonable to initially set interval to time.Duration(math.MaxInt64); because before "update" is called, after "stdtime.After(interval)" return， it does nothing meaningful。 In the processing of the "update" event, the initial interval will be overwritten by params.RunInterval.

joestringer commented on pkg/controller/controller.go:
```diff
@@ -239,93 +241,100 @@ func (c *controller) GetLastErrorTimestamp() time.Time {
-				// next controller update (or stop).
-				err = NewExitReason("controller context canceled")
-			}
+		if runFunc && params.DoFunc != nil {
```
Why would `DoFunc` be nil?

What I would be concerned about here is if the code silently stops running a controller because some code unintentionally cleared the DoFunc. If we are going to check for this condition, it should probably instead record a special error to indicate this condition. Alternatively, don't check for this.

joestringer replied:
This seems related to the comment below: https://github.com/cilium/cilium/pull/38031#discussion_r1989678322 . That is, params.DoFunc might be nil because the controller was triggered before the params were received.

I think it should be possible to write this in a way that the controller has the parameters first before running so that this code doesn't have to check for the `params.DoFunc != nil`.

orange30 replied:
> that the controller has the parameter

I will add a log to record the special error.

orange30 replied:
> This seems related to the comment below: [#38031 (comment)](https://github.com/cilium/cilium/pull/38031#discussion_r1989678322) . That is, params.DoFunc might be nil because the controller was triggered before the params were received.

yes.

> I think it should be possible to write this in a way that the controller has the parameters first before running so that this code doesn't have to check for the `params.DoFunc != nil`.

Here we introduce the delay,  i think we can not ensure the triggered is after the "update".

joestringer replied:
I think we should enforce that `params.DoFunc` is never nil. Why is it nil?

joestringer commented on pkg/controller/controller.go:
```diff
@@ -239,93 +241,100 @@ func (c *controller) GetLastErrorTimestamp() time.Time {
 			// update channel is never closed
+			runFunc = true
 		case <-stdtime.After(interval):
 			// timer channel is not yet closed
```
Should the `stdtime.After(...)` case also set runFunc?

joestringer replied:
See the danger here: https://github.com/cilium/cilium/pull/38031#discussion_r1989678322

If this also sets runFunc then `params` might be empty.

orange30 replied:
> Should the `stdtime.After(...)` case also set runFunc?

Only update and trigger should set runFunc to be true.
"update" mean that "params" has been setted correctly, controller should execute the params.DoFunc now.

"trigger" mean it shoud exectued immediately:
(1) If "update" has been executed before, “params” should have been configured correctly, for example params.DoFunc will be setted correctly.
(2) "params" is a local variable in this function and will never be nil.
(3) If "update" has not been executed before, (here we introduced delay, this situation is possible), we need to be compatible with the situation where params is not set correctly( in this case : before "update" is called, params.DoFunc will always be nil.)

I think the introduction of "stdtime.After(interval)" is more to make the controller compatible with the situation where params.DoFunc is executed periodically.
(1) For the situation where "controller does not need to be executed periodically", params.RunInterval can be left unset or set to 0.
    // If no run interval is specified, no further updates
    // are required.
    if interval == time.Duration(0) {
        // Don't exit the goroutine, since that only happens when the
        // controller is explicitly stopped. Instead, just wait for
        // the next update. 
        c.getLogger().Debug("Controller run succeeded; waiting for next controller update or stop")
        interval = time.Duration(math.MaxInt64)
    }
(2) For the situation where "controller needs to be executed periodically", params.RunInterval needs to be set to a non-zero value. Even in this case, params.DoFunc should be called during the "update" event.

joestringer replied:
> (3) If "update" has not been executed before, (here we introduced delay, this situation is possible), we need to be compatible with the situation where params is not set correctly( in this case : before "update" is called, params.DoFunc will always be nil.)

I'm suggesting that we don't allow this order of operations. The controller is declared with parameters, so those parameters should be available right from the initial declaration. It doesn't make sense to run a controller without using its parameters.

joestringer replied:
Strictly speaking, we do not need to enable this boolean in the `stdtime.After(interval)` case. I think that if we ensure that the parameters are available right from the start, we won't need to make that change and we won't need the `runFunc` bool at all either.

orange30 replied:
Yes. Please help review it again.

joestringer commented on pkg/controller/manager.go:
```diff
@@ -52,6 +52,15 @@ func (m *Manager) UpdateController(name string, params ControllerParams) {
+	select {
+	case <-time.After(params.DelayUpdate):
+		ctrl.update <- params
+	case <-ctrl.terminated:
```
nit: The first line here could block, should there be another `select {}` statement to handle the `<-ctrl.terminated` case after the delay?

orange30 replied:
This select is not in the loop，I think we donot need another select {} statement to handle the <-ctrl.terminated. 
What we expect is that:
time.After or ctrl.terminated: the one that returns first will be executed, otherwise it should be blocked here.

joestringer replied:
Consider this case: `After()` was called and the goroutine made no progress for some undefined amount of time before proceeding to execute the `ctrl.update <- params` line. 
The `runController()` code that receives from that channel could have received `<-c.stop` immediately before the `After()` returns, then `runController()` proceeds to call into `params.StopFunc(...)` but not yet called `close(c.terminated)`. In this case, this goroutine would be stuck on `ctrl.update <- params` forever because the goroutine which normally reads from that channel will never read from it again.

joestringer commented on pkg/defaults/defaults.go:
```diff
@@ -166,6 +166,9 @@ const (
 	// option.IdentityChangeGracePeriod
 	IdentityChangeGracePeriod = 5 * time.Second
 
+	// Delay interval of CiliumIdentidy after Namespace lable changed..
```
nit: 
```suggestion
	// Delay to change CiliumIdentity after Namespace label changed.
```

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -445,6 +445,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change")
 	option.BindEnv(vp, option.IdentityChangeGracePeriod)
 
+	flags.Duration(option.DelayCIdentityOfNsLabelChange, 30*time.Second, "Delay interval of CiliumIdentidy after Namespace lable changed.")
```
```suggestion
	flags.Duration(option.DelayCIdentityOfNsLabelChange, defaults.DelayCIdentityOfNsLabelChange, "Randomly delay CiliumIdentity updates for Endpoints by up to this time after Namespace labels change")
```

joestringer commented on pkg/option/config.go:
```diff
@@ -1799,6 +1802,9 @@ type DaemonConfig struct {
 	// to whitelist the new upcoming identity of the endpoint.
 	IdentityChangeGracePeriod time.Duration
 
+	// Delay interval of CiliumIdentidy after Namespace lable changed..
```
```suggestion
	// Delay to change CiliumIdentity after Namespace label changed.
```

joestringer commented on pkg/controller/controller.go:
```diff
@@ -239,93 +241,100 @@ func (c *controller) GetLastErrorTimestamp() time.Time {
 	errorRetries := 1
+	runFunc := false
+	interval := 10 * time.Minute
 
```
nit: If the intent is to wait for the first trigger / params update, maybe instead of changing the loop the code should just wait for that update here?

```suggestion
	select {
		case <-c.stop:
			goto shutdown

		case params = <-c.update:
			// update channel is never closed
			runFunc = true
		case <-stdtime.After(interval):
			// timer channel is not yet closed
		case <-c.trigger:
			// trigger channel is never closed
			runFunc = true
	}
```

But wait, then the flaw is revealed: if `c.trigger()` happens first then `params` is empty, which is unacceptable. This also needs fixing.

orange30 replied:
"params" is a local variable in this function and will never be nil.
https://github.com/cilium/cilium/pull/38031/files#r1991017071

joestringer replied:
Sorry, I misspoke. I meant the variable would be uninitialized rather than nil. My point is that we should not be using the values from an uninitialized structure.

joestringer commented on pkg/controller/manager.go:
```diff
@@ -52,6 +52,20 @@ func (m *Manager) UpdateController(name string, params ControllerParams) {
+		select { // Refer to func (m *Manager) createControllerLocked
+		case ctrl.update <- params:
+		default:
+		}
```
Could this lose updates if `runController()` is not ready to receive the update event?

Maybe this is more robust?
```suggestion
		// Attempt to send the parameters update unless there is a race to shut down.  
		select {
		case ctrl.update <- params:
		case <-ctrl.stop:
		}
```

joestringer commented on pkg/controller/manager.go:
```diff
@@ -74,9 +88,13 @@ func (m *Manager) updateController(name string, params ControllerParams) *manage
+			select {
+			case ctrl.update <- ctrl.params:
+			default:
+			}
```
Hmm interesting, this doesn't seem right to me either. I appreciate this is how the existing code already worked, but it seems to me like if the controller is not ready to receive an update (because for instance it is actively processing its run function), then this update will just be dropped. That seems bad also. Maybe we aren't using `UpdateController()` that often after startup so we never noticed this problem before.

At the risk of causing backpressure up the callstack, this seems more robust:
```suggestion
			// Attempt to send the parameters update unless there is a race to shut down.
			select {
			case ctrl.update <- ctrl.params:
			case <-ctrl.stop
			}
```

joestringer replied:
Apparently there is a whole thread about this bug in the tree already :sweat_smile: https://github.com/cilium/cilium/issues/29015

orange30 replied:
> Hmm interesting, this doesn't seem right to me either. I appreciate this is how the existing code already worked, but it seems to me like if the controller is not ready to receive an update (because for instance it is actively processing its run function), then this update will just be dropped. That seems bad also. Maybe we aren't using `UpdateController()` that often after startup so we never noticed this problem before.
> 

Yes, I agress that "maybe we aren't using UpdateController() that often after startup".

> At the risk of causing backpressure up the callstack, this seems more robust:

1. "Backpressure up the callstack" seems to have some risks that need to assess.
2. In func (m *Manager) createControllerLocked, when we initialize the update channel, why not we initialize the channel size to be 32?
3. If we can set the update channel size to be 32, and then we use this change, then I think the risk of "Backpressure up the callstack" will be Insignificant。

joestringer commented on pkg/controller/manager.go:
```diff
@@ -108,6 +126,16 @@ func (m *Manager) createControllerLocked(name string, params ControllerParams) *
+		go m.delayUpdate(ctrl, ctrl.params)
+	}
+
 	go ctrl.runController(ctrl.params)
```
Isn't the first update lost because `go ctrl.runController(...)` isn't run until after? So there's nothing to receive/handle the message on the other side, meaning the `select {...}` statement becomes a non-blocking send?

joestringer replied:
For the `DelayUpdate == 0` case I think we can just move the `go ctrl.runController()` above this hunk and then perform a blocking send. Then for the delay update case I think it's already doing a blocking send so it should be good to go.

orange30 replied:
Now the size of ctrl.update channel is 1, so I think the first "ctrl.update <- ctrl.params" will success. 
"move the go ctrl.runController() above this hunk" looks fine, I will fix it.

sypakine commented on daemon/cmd/daemon_main.go:
```diff
@@ -445,6 +445,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change")
 	option.BindEnv(vp, option.IdentityChangeGracePeriod)
 
+	flags.Duration(option.DelayCIdentityOfNsLabelChange, 30*time.Second, "Randomly delay CiliumIdentity updates for Endpoints by up to this time after Namespace labels change")
```
The more accurate and concise term is jitter:

https://en.wikipedia.org/wiki/Jitter

sypakine replied:
nit: `DelayCIdentityOfNsLabelChange` is confusing and too narrowly specified. Something like `CiliumIdentityMaxJitter` can be extended to be used elsewhere if necessary.

sypakine commented on pkg/controller/controller.go:
```diff
@@ -105,6 +105,9 @@ type ControllerParams struct {
 
 	Context context.Context
+
+	// The interval for which this update needs to be delayed
```
nit: 

```
// UpdateJitter is the duration to delay processing the update event.
```

sypakine commented on daemon/cmd/daemon_main.go:
```diff
@@ -445,6 +445,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change")
 	option.BindEnv(vp, option.IdentityChangeGracePeriod)
 
+	flags.Duration(option.CiliumIdentityMaxJitter, 30*time.Second, "Maximum jitter time for CiliumIdentity")
```
reference the default value:

```
flags.Duration(option.CiliumIdentityMaxJitter, defaults.CiliumIdentityMaxJitter, "Maximum jitter time for CiliumIdentity")
```

sypakine commented on pkg/defaults/defaults.go:
```diff
@@ -166,6 +166,9 @@ const (
 	// option.IdentityChangeGracePeriod
 	IdentityChangeGracePeriod = 5 * time.Second
 
+	//Maximum jitter time for CiliumIdentity
```
nit: 

```
// CiliumIdentityMaxJitter is the maximum duration to delay processing a CiliumIdentity under certain conditions.
```

sypakine commented on pkg/endpoint/endpoint.go:
```diff
@@ -1912,6 +1912,14 @@ func (e *Endpoint) RunRestoredMetadataResolver(bwm dptypes.BandwidthManager, res
+	return e.modifyIdentityLabelsWithJitter(source, addLabels, delLabels, 0)
+}
+
+func (e *Endpoint) ModifyIdentityLabelsWithJitter(source string, addLabels, delLabels labels.Labels, UpdateJitter time.Duration) error {
```
I'll defer to Isovalent folks but I'm wary of this pattern as it may lead to proliferation of methods.

Options struct or functional options pattern can keep this in check.

joestringer replied:
Should be OK to add the parameter and update the callsites rather than adding a wrapper?

```
$ git grep ModifyIdentityLabels
daemon/cmd/endpoint.go: if err := ep.ModifyIdentityLabels(labels.LabelSourceAny, addLabels, delLabels); err != nil {
pkg/endpoint/endpoint.go:// ModifyIdentityLabels changes the custom and orchestration identity labels of an endpoint.
pkg/endpoint/endpoint.go:func (e *Endpoint) ModifyIdentityLabels(source string, addLabels, delLabels labels.Labels) error {
pkg/endpoint/endpoint.go:       changed, err := e.OpLabels.ModifyIdentityLabels(addLabels, delLabels)
pkg/endpoint/endpoint.go:       err := e.ModifyIdentityLabels(source, newIdtyLabels, oldIdtyLabels)
pkg/k8s/watchers/namespace.go:                  err := ep.ModifyIdentityLabels(labels.LabelSourceK8s, newIdtyLabels, oldIdtyLabels)
pkg/labels/oplabels.go:func (o *OpLabels) ModifyIdentityLabels(addLabels, delLabels Labels) (changed bool, err error) {
```

sypakine commented on pkg/k8s/watchers/namespace.go:
```diff
@@ -122,6 +125,13 @@ func getNamespaceLabels(ns *slim_corev1.Namespace) labels.Labels {
 
+// randomDuration generates a random time.Duration between 0 and maxDuration
+func randomDuration(maxDuration time.Duration) time.Duration {
+	// Generate a random duration
```
nit: omit comment as this is self-evident:

```
	return time.Duration(rand.Int64N(int64(maxDuration)))
```

sypakine commented on pkg/option/config.go:
```diff
@@ -825,6 +825,9 @@ const (
 	// IdentityChangeGracePeriod option
 	IdentityChangeGracePeriod = "identity-change-grace-period"
 
+	// Maximum jitter time for CiliumIdentity (default 30s)
```
nit:

```
// CiliumIdentityMaxJitter is the maximum duration to delay processing a CiliumIdentity under certain conditions (default: 30s).
```

sypakine commented on pkg/k8s/watchers/namespace.go:
```diff
@@ -139,7 +149,8 @@ func (u *namespaceUpdater) update(newNS *slim_corev1.Namespace) error {
 		if newNS.Name == epNS {
-			err := ep.ModifyIdentityLabels(labels.LabelSourceK8s, newIdtyLabels, oldIdtyLabels)
+			err := ep.ModifyIdentityLabelsWithJitter(labels.LabelSourceK8s, newIdtyLabels, oldIdtyLabels,
+				randomDuration(option.Config.CiliumIdentityMaxJitter))
```
consider: set as a field in the namespaceUpdater rather than referencing the global config

sypakine reviewed (commented):
Could you add a sentence or two to each commit message body? For the first, can use `Related #38030`

sypakine commented on pkg/endpoint/endpoint.go:
```diff
@@ -1911,7 +1911,7 @@ func (e *Endpoint) RunRestoredMetadataResolver(bwm dptypes.BandwidthManager, res
 // endpoint will receive a new identity and will be regenerated. Both of these
 // operations will happen in the background.
-func (e *Endpoint) ModifyIdentityLabels(source string, addLabels, delLabels labels.Labels) error {
+func (e *Endpoint) ModifyIdentityLabels(source string, addLabels, delLabels labels.Labels, UpdateJitter time.Duration) error {
```
`updateJitter`

sypakine commented on pkg/endpoint/endpoint.go:
```diff
@@ -2097,7 +2097,7 @@ func (e *Endpoint) identityResolutionIsObsolete(myChangeRev int) bool {
 //
 // Must be called with e.mutex NOT held.
-func (e *Endpoint) runIdentityResolver(ctx context.Context, blocking bool) (regenTriggered bool) {
+func (e *Endpoint) runIdentityResolver(ctx context.Context, blocking bool, UpdateJitter time.Duration) (regenTriggered bool) {
```
`updateJitter`

sypakine reviewed (commented):
[Point 5](https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#submitting-a-pull-request) states the commit should have a "well written commit description".

Can you reference an integration/e2e test that exercises namespace label changes? Want to make sure coverage for this exists.

sypakine reviewed (commented):
LGTM; running tests before approving.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -445,6 +445,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change")
 	option.BindEnv(vp, option.IdentityChangeGracePeriod)
 
+	flags.Duration(option.CiliumIdentityMaxJitter, defaults.CiliumIdentityMaxJitter, "Maximum jitter time for CiliumIdentity")
```
```suggestion
	flags.Duration(option.CiliumIdentityMaxJitter, defaults.CiliumIdentityMaxJitter, "Maximum jitter time to begin processing CiliumIdentity updates")
```

joestringer commented on pkg/controller/manager.go:
```diff
@@ -94,7 +112,7 @@ func (m *Manager) createControllerLocked(name string, params ControllerParams) *
 			uuid:       uuid.New().String(),
 			stop:       make(chan struct{}),
-			update:     make(chan ControllerParams, 1),
+			update:     make(chan ControllerParams, 32),
```
Why does this channel capacity need to change?

The concern I have is that we *must* make this code robust to blocking sends on the channel. Channel size is primarily useful as a buffer, not to handle concurrency issues. I don't think we expect to see many controller params updates, so maybe this should just stay at capacity 1?

orange30 replied:
Ok, I change it to be 1.

joestringer commented on pkg/endpoint/api.go:
```diff
@@ -576,7 +576,7 @@ func (e *Endpoint) ProcessChangeRequest(newEp *Endpoint, validPatchTransitionSta
 	if rev != 0 {
 		// Run as a goroutine since the runIdentityResolver needs to get the lock
-		go e.runIdentityResolver(e.aliveCtx, false)
+		go e.runIdentityResolver(e.aliveCtx, false, 0)
```
I recognize that the goal of this PR is to mitigate namespace updates, but I wonder whether these other cases that are currently passing `0` should also have the new jitter behavior?

orange30 replied:
> I recognize that the goal of this PR is to mitigate namespace updates, but I wonder whether these other cases that are currently passing `0` should also have the new jitter behavior?

They should not have jitter behavior.

joestringer replied:
What is your thinking behind saying that the other cases should not have jitter behavior? Is there something fundamentally different about those other updates?

orange30 replied:
1. From the caller's perspective: UpdateController's comment says: "Updating a controller will cause the DoFunc to be run immediately regardless of any previous conditions." Before we added the UpdateJitter parameter, when callers designed the logic related to calling functions, it was estimated that most of them assumed that UpdateController would execute DoFunc immediately. (Although the previous logic could not 100% guarantee immediate execution)
2. From the controller's perspective: When UpdateController is called, it is more reasonable if UpdateJitter is an absolute value: when UpdateJitter is 0, it intuitively means that it should be executed directly without delay.
3. From the perspective of patch modifying without introducing new risks: If we try to keep consistent with the previously declared logic, the probability of introducing unexpected new problems will be smaller.

joestringer replied:
I understand the concern that changing these other paths might incur additional risk, but this is a proposed change for v1.18 (ETA mid 2025) so we have a few months to test and validate the change. However, I am looking for reasoning about the logically correct path, not just what minimizes risk.

This specific path mentioned on this thread handles a PATCH event for the endpoint object on the REST API. This makes it very unlikely it'll cause the sorts of problems being addressed by this PR, so I think it is fine to keep this with zero jitter.

joestringer commented on pkg/controller/controller.go:
Looking afresh at the first patch, I'm wondering whether the extra goroutines and channels to delay updates are more complicated than we really need to solve the core problem. Why don't we just delay the `DoFunc()` within the existing core `runController()` loop? I'm thinking something like this to replace the first patch:

```
diff --git a/pkg/controller/controller.go b/pkg/controller/controller.go
index 0283b3247bc4..0722c6f61756 100644
--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -8,6 +8,7 @@ import (
        "errors"
        "fmt"
        "math"
+       "math/rand/v2"
        stdtime "time"
 
        "github.com/cilium/hive/cell"
@@ -105,6 +106,9 @@ type ControllerParams struct {
        NoErrorRetry bool
 
        Context context.Context
+
+       // Jitter is the duration to delay executing DoFunc.
+       Jitter time.Duration
 }
 
 // undefinedDoFunc is used when no DoFunc is set. controller.DoFunc is set to this
@@ -248,12 +252,17 @@ func (c *controller) runController(params ControllerParams) {
                interval := params.RunInterval
 
                start := time.Now()
+               jitter := time.Duration(0)
+               if params.Jitter > 0 {
+                       jitter = time.Duration(rand.Int64N(int64(params.Jitter)))
+               }
+               time.Sleep(jitter)
                err = params.DoFunc(params.Context)
                duration := time.Since(start)
 
                c.mutex.Lock()
                c.lastDuration = duration
-               c.getLogger().Debug("Controller func execution time: ", c.lastDuration)
+               c.getLogger().Debug("Controller func executed", "duration", c.lastDuration, "jitter", jitter)
 
                if err != nil {
                        if params.Context.Err() != nil {
```

joestringer replied:
If we do this then `randomDuration()` could be removed from the k8s/watchers changes since the randomness is calculated in the inner loop.

orange30 replied:
> Looking afresh at the first patch, I'm wondering whether the extra goroutines and channels to delay updates are more complicated than we really need to solve the core problem. Why don't we just delay the `DoFunc()` within the existing core `runController()` loop? I'm thinking something like this to replace the first patch:
> 

If we use this patch, some update events may be lost。
Consider the following scenarios:
1. First of all, we modify the namespace label, the call stack is : func (u *namespaceUpdater) update() -> func (e *Endpoint) ModifyIdentityLabels () -> func (e *Endpoint) runIdentityResolver() ->e.controllers.UpdateController (..., updateJitter), for example, updateJitter is 25s.
2. Then, the label of the pod changes, and the call stack is: ...-> func (e *Endpoint) ModifyIdentityLabels () -> func (e *Endpoint) runIdentityResolver() ->e.controllers.UpdateController (..., 0) .
Now because the func (c *controller) runController is hanged for time.sleep(), ctrl.update channel will saves ctrl.params.
3. If the pod’s label changes again, then ctrl.params will cannot be saved to ctrl.update because the length of ctrl.update is 1. Then this update will be lost forever.
https://github.com/cilium/cilium/blob/77552f994d596891168741e903c63a566a15cd80/pkg/controller/manager.go#L83
// Notify the goroutine of the params update.
select {
case ctrl.update <- ctrl.params:
default:
}

This scenario should not happen too often, but it is unacceptable to cause update events to be lost.
If we initialize the length of ctrl.update to 32, it seems that this problem could be solved.
Another benefit of this change is that it linearizes the processing of multiple update events received by a specific controller.

joestringer replied:
Thanks for the analysis. I think that the controller is being a bit misused in this case. Controller is intended as an async loop that can be triggered at some point in the future, it is *not* intended to be a lossless queue. Related, I've proposed a change that would ensure that eventually the configuration which is applied to the controller will take effect: https://github.com/cilium/cilium/pull/38853 .

For (2), it should use the same configuration for the updateJitter as we discussed in [this thread](https://github.com/cilium/cilium/pull/38031#discussion_r2034019587). Technically there are other ways to trigger a similar situation using the local REST API, but it is rare to use that API and I'm not sure it's worth building a more complicated solution just to address that particular case.

With #38853 and the proposed solution above I think that we should no longer lose important configuration updates.

orange30 replied:
https://github.com/cilium/cilium/issues/29015#issuecomment-1822600339
It looks #38853 is a better solution, after it is merged, I will rebase this pr again.

joestringer commented on daemon/cmd/endpoint.go:
```diff
@@ -1109,7 +1109,7 @@ func (d *Daemon) modifyEndpointIdentityLabelsFromAPI(id string, add, del labels.
 	}
 
-	if err := ep.ModifyIdentityLabels(labels.LabelSourceAny, addLabels, delLabels); err != nil {
+	if err := ep.ModifyIdentityLabels(labels.LabelSourceAny, addLabels, delLabels, 0); err != nil {
```
I think it's fair to have no jitter on local REST API requests :+1:

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2063,7 +2063,7 @@ func (e *Endpoint) UpdateLabels(ctx context.Context, sourceFilter string, identi
 	e.unlock()
 	if rev != 0 {
-		return e.runIdentityResolver(ctx, blocking)
+		return e.runIdentityResolver(ctx, blocking, 0)
```
It's not super obvious but tracing up the incoming calls to here reaches either local reserved endpoint creation (health, host, ingress); endpoint restoration on startup; or brand new endpoint creation, none of which require jitter to mitigate the issue outlined in the PR description. So I agree `0` is the right jitter here.

```
INCOMING CALLS                                                                                                                                                   
- f UpdateLabels github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                             
  - f LaunchAsEndpoint github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                       
    + f initHealth github.com/cilium/cilium/cilium-health/launch • health.go                                                                                     
  - f createEndpoint github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                         
    + f putEndpointIDHandler github.com/cilium/cilium/daemon/cmd • endpoint.go                                                                                   
  - f metadataResolver github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                       
    - f RunMetadataResolver github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                  
      + f createEndpoint github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                     
      - f RunRestoredMetadataResolver github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                        
        - f validateEndpoint github.com/cilium/cilium/pkg/endpoint • state.go                                                                                    
          + f restoreOldEndpoints github.com/cilium/cilium/daemon/cmd • state.go                                                                                 
        + f RegenerateAfterRestore github.com/cilium/cilium/pkg/endpoint • restore.go                                                                                                                                                     
  - f InitWithIngressLabels github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                  
    - f AddIngressEndpoint github.com/cilium/cilium/pkg/endpoint • manager.go                                                                                    
      + f startDaemon github.com/cilium/cilium/pkg/endpointmanager • daemon_main.go                                                                              
  - f InitWithNodeLabels github.com/cilium/cilium/pkg/endpoint • endpoint.go                                                                                     
    - f initHostEndpointLabels github.com/cilium/cilium/pkg/endpoint • manager.go                                                                                
      + f AddHostEndpoint github.com/cilium/cilium/pkg/endpointmanager • manager.go                                                                              
      + f InitHostEndpointLabels github.com/cilium/cilium/pkg/endpointmanager • manager.go
```

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2077,7 +2077,7 @@ func (e *Endpoint) UpdateLabelsFrom(oldLbls, newLbls map[string]string, source s
 	oldIdtyLabels, _ := labelsfilter.Filter(oldLabels)
 
-	err := e.ModifyIdentityLabels(source, newIdtyLabels, oldIdtyLabels)
+	err := e.ModifyIdentityLabels(source, newIdtyLabels, oldIdtyLabels, 0)
```
This case traces up to `updateK8sPodV1` which could plausibly have a similar problem to the namespaces problem you mentioned. I think that for this case it would also make sense to configure `option.Config.CiliumIdentityMaxJitter`

orange30 replied:
ok, I will modify it in next rebase.

joestringer reviewed (changes requested):
Apologies about the delay, I have been traveling recently and haven't dedicated the time to thoroughly review the change.

I dug back through all the paths for configuring jitter and identified (a) A potential simpler implementation to both solve the current unit test failure as well as avoiding any extra goroutines or channel coordination; and (b) Another path from k8s that I think should have this jitter applied to it. See my comments below.

With these changes and a rebase against the latest `main` I'm hoping that all the tests should pass.

joestringer reviewed (approved):
Looks good to me. Thanks for your patience as we worked through the review process to try to simplify the implementation to make it more reliable.

sypakine reviewed (approved):
Apologies for the delay, I just returned for 3 weeks of OOO with no github access.

orange30 committed:
controller: support update jitter

A Jitter parameter has been added to the controller.
When Jitter is set to 0, the params.DoFunc is executed immediately.
Otherwise, it will pause for a random period between 0 and Jitter before executing the function.

Fixed: #38030
Signed-off-by: zhikuodu <duzhk@qq.com>

orange30 committed:
scale: import identity update jitter after namespace label change

When namespace label is changed, the identity update incorporates a jitter to manage the update timing.
The maximum jitter time for CiliumIdentity can be configured using the "identity-max-jitter" key in the cilium-config.

Fixes: #38030
Signed-off-by: zhikuodu <duzhk@qq.com>

gandro commented on pkg/controller/controller.go:
```diff
@@ -312,6 +316,11 @@ func (c *controller) runController() {
+		jitter := time.Duration(0)
+		if params.Jitter > 0 {
+			jitter = time.Duration(rand.Int64N(int64(params.Jitter)))
+			time.Sleep(jitter)
```
nit: Since the sleep is potentially quite long, would it make sense to use an interruptible sleep here (i.e. a select statement that waits both on `time.After` and on `params.Context.Done()` ?

orange30 replied:
@gandro Thanks, I think this is a good idea! 
Maybe the change looks like:
<img width="792" alt="Clipboard_Screenshot_1749651280" src="https://github.com/user-attachments/assets/e3af4901-a2ec-49a3-928f-c871ff725b6d" />
 @joestringer Do you have any suggestions?

gandro replied:
Yeah this looks good to me!

gandro reviewed (approved):
Seems pretty straight forward
